pennylane/transforms/hamiltonian_expand.py

```diff
@@ -191,3 +196,183 @@ def processing_fn(res):
+            qml.Hadamard(wires=0)
+            qml.CNOT(wires=[0, 1])
+            qml.PauliX(wires=2)
+            qml.expval(S)
+
+    With grouping, the Sum gets split into two groups of observables (here
+    ``[qml.PauliZ(0), qml.s_prod(2, qml.PauliX(1))]`` and ``[qml.s_prod(3, qml.PauliX(0))]``):
+
+    >>> tapes, fn = qml.transforms.sum_expand(tape, group=True)
+    >>> for tape in tapes:
+    ...     print(tape.measurements)
+    [expval(PauliZ(wires=[0])), expval(2*(PauliX(wires=[1])))]
+    [expval(3*(PauliX(wires=[0])))]
+    """
+    non_expanded_measurements = []
+    non_expanded_measurement_idxs = []
+    expanded_tapes = []
+    expanded_measurement_idxs = []
+    num_tapes = []
+    for idx, m in enumerate(tape.measurements):
+        if isinstance(m.obs, Sum) and m.return_type is Expectation:
+            sum_op = m.obs
+            if group:
+                # make one tape per group of qwc observables
+                obs_groupings = _group_summands(sum_op)
+                tapes = [
+                    QuantumScript(tape._ops, [qml.expval(o) for o in obs], tape._prep)
+                    for obs in obs_groupings
+                ]
+            else:
+                # make one tape per summand
+                tapes = [
+                    QuantumScript(ops=tape.operations, measurements=[qml.expval(summand)])
+                    for summand in sum_op.operands
+                ]
+
+            expanded_tapes.extend(tapes)
+            expanded_measurement_idxs.append(idx)
+            num_tapes.append(len(tapes))
+
+        else:
+            non_expanded_measurements.append(m)
+            non_expanded_measurement_idxs.append(idx)
+
+    non_expanded_tape = (
+        [QuantumScript(ops=tape._ops, measurements=non_expanded_measurements, prep=tape._prep)]
+        if non_expanded_measurements
+        else []
+    )
+    tapes = expanded_tapes + non_expanded_tape
+    # Indices used for reordering the obtained results in the postprocessing function
+    measurement_idxs = expanded_measurement_idxs + non_expanded_measurement_idxs
+
+    def processing_fn(res):
+        processed_results = []
+        # process results of all tapes except the last one
+        for idx, n_tapes in enumerate(num_tapes):
+            sum_results = res[idx : idx + n_tapes]
+            processed_results.extend([_process_sum_results(sum_results, group=group)])
+        # add results of last tape, which contains all the non-sum observables
+        if non_expanded_tape:
+            non_expanded_res = [res[-1]] if len(non_expanded_measurement_idxs) == 1 else res[-1]
+            processed_results.extend(non_expanded_res)
+        # sort results
+        sorted_results = sorted(zip(processed_results, measurement_idxs), key=lambda x: x[1])
+        if len(sorted_results) > 1:
+            return [res[0] for res in sorted_results]
+        return sorted_results[0][0]
+
+    return tapes, processing_fn
+
+
+def _group_summands(sum: Sum):
```

Comment:
Ah wow! Is there a reason why you wrote this from scratch and not `qml.grouping.group_observables` (or the new version that Jay is introducing for Pauli arithmetics)? TBH I cant tell which one is better or if there even is a difference at all. Can you? https://docs.pennylane.ai/en/stable/_modules/pennylane/grouping/group_observables.html

Reply:
This should be replaced with the new version of `group_observables` that @Jaybsoni is working on.

I can add a #2967 #TODO message

The method defined here is very simple: it only takes into account operator wires